File: libs/rudder-integration-sprig-react-native/ios/RNRudderSprigIntegrationModuleImpl.m
#import "RNRudderSprigIntegrationModuleImpl.h"
#import <UIKit/UIKit.h>
#import <RNRudderSdk/RNRudderAnalytics.h>

// Forward-declare the Swift `RudderSprigFactory` rather than importing
// <Rudder_Sprig/Rudder_Sprig-Swift.h>. The generated Swift header transitively
// pulls in the Rudder umbrella, which defines `NSString *const SDK_VERSION`
// inside RSVersion.h — a non-extern global that produces a duplicate-symbol
// link error against libRudder.a's RSConstants.o.
//
// The objc_runtime_name attribute must match the mangled name Swift emits
// (`_TtC12Rudder_Sprig18RudderSprigFactory`) so the ObjC linker resolves
// `[RudderSprigFactory instance]` to the Swift class symbol.
__attribute__((objc_runtime_name("_TtC12Rudder_Sprig18RudderSprigFactory")))
@interface RudderSprigFactory : NSObject <RSIntegrationFactory>
+ (RudderSprigFactory *)instance;
- (void)setViewController:(UIViewController *)viewController;
@end

@implementation RNRudderSprigIntegrationModuleImpl

- (void)setup:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject {
    RudderSprigFactory *factory = [RudderSprigFactory instance];
    [RNRudderAnalytics addIntegration:factory];

    // Provide the host app's root view controller so trackAndPresent surveys can show.
    dispatch_async(dispatch_get_main_queue(), ^{
        UIViewController *rootVC = [self currentRootViewController];
        if (rootVC != nil) {
            [factory setViewController:rootVC];
        }
    });

    resolve(nil);
}

- (UIViewController *)currentRootViewController {
    UIWindow *keyWindow = nil;
    NSSet<UIScene *> *scenes = [[UIApplication sharedApplication] connectedScenes];
    for (UIScene *scene in scenes) {
        if (scene.activationState != UISceneActivationStateForegroundActive) {
            continue;
        }
        if (![scene isKindOfClass:[UIWindowScene class]]) {
            continue;
        }
        UIWindowScene *windowScene = (UIWindowScene *)scene;
        for (UIWindow *window in windowScene.windows) {
            if (window.isKeyWindow) {
                keyWindow = window;
                break;
            }
        }
        if (keyWindow != nil) {
            break;
        }
    }
    return keyWindow.rootViewController;
}

@end

File: libs/rudder-integration-sprig-react-native/ios/RNRudderSprigIntegrationModuleImpl.swift
import Foundation
import UIKit
import React
import RNRudderSdk
import Rudder
import Rudder_Sprig

@objc(RNRudderSprigIntegrationModuleImpl)
public final class RNRudderSprigIntegrationModuleImpl: NSObject {
    private var integrationRegistered = false

    @objc public func setup(_ resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
        let factory = RudderSprigFactory.instance

        // Guard against re-registration on hot reload / repeated setup() calls
        // (mirrors the Android `callbacksRegistered` guard in the Java impl).
        if !integrationRegistered {
            RNRudderAnalytics.addIntegration(factory)
            integrationRegistered = true
        }

        // Provide the host app's root view controller so trackAndPresent surveys can show.
        // Resolve only after the VC is wired, so JS callers awaiting setup() can safely
        // fire survey-triggering events on the next tick.
        DispatchQueue.main.async {
            if let rootVC = self.currentRootViewController() {
                factory.setViewController(rootVC)
            }
            resolve(nil)
        }
    }

    private func currentRootViewController() -> UIViewController? {
        var keyWindow: UIWindow?
        for scene in UIApplication.shared.connectedScenes {
            guard scene.activationState == .foregroundActive,
                  let windowScene = scene as? UIWindowScene else { continue }
            if let found = windowScene.windows.first(where: { $0.isKeyWindow }) {
                keyWindow = found
                break
            }
        }

        // Walk the presentation chain so we hand Sprig the topmost VC. If the host
        // app already has a modal up (login sheet, paywall, etc.), presenting the
        // survey from rootViewController would throw "already presenting".
        var vc = keyWindow?.rootViewController
        while let presented = vc?.presentedViewController {
            vc = presented
        }
        return vc
    }
}
